receiver/mongodbreceiver/client.go

```diff
@@ -0,0 +1,222 @@
+	return c.RunCommand(ctx, database, bson.M{"serverStatus": 1})
+}
+
+// DBStats returns the result of db.runCommand({ dbStats: 1 })
+// more information can be found here: https://docs.mongodb.com/manual/reference/command/dbStats/
+func (c *mongodbClient) DBStats(ctx context.Context, database string) (bson.M, error) {
+	return c.RunCommand(ctx, database, bson.M{"dbStats": 1})
+}
+
+type mongoTestConnectionResponse struct {
+	Version string
+}
+
+// TestConnection validates the underlying connection of the client and returns a result of the
+// version of mongo the client is connected to so adjustments in collection protocol can
+// be determined
+func (c *mongodbClient) TestConnection(ctx context.Context) (*mongoTestConnectionResponse, error) {
+	if err := c.Connect(ctx); err != nil {
+		return nil, fmt.Errorf("unable to connect: %w", err)
+	}
+
+	res, err := c.RunCommand(ctx, "admin", bson.M{"buildInfo": 1})
+	if err != nil {
+		return nil, fmt.Errorf("unable to get build info: %w", err)
+	}
+
+	v, ok := res["version"].(string)
+	if !ok {
+		return nil, errors.New("unable to parse mongo version from server")
+	}
+
+	c.logger.Debug(fmt.Sprintf("retrieved version information from mongo server to be: %s", v))
+	return &mongoTestConnectionResponse{
+		Version: v,
+	}, nil
+}
+
+func (c *mongodbClient) ensureClient(ctx context.Context) error {
+	if !c.isConnected(ctx) {
+		driver, err := mongo.Connect(ctx, c.options())
+		if err != nil {
+			return fmt.Errorf("error creating mongo client: %w", err)
+		}
+
+		if err := driver.Ping(ctx, readpref.PrimaryPreferred()); err != nil {
+			return fmt.Errorf("could not connect to hosts %v: %w", c.hosts, err)
+		}
+
+		c.logger.Info(fmt.Sprintf("Mongo connection established to hosts: %s", strings.Join(c.hosts, ", ")))
+		c.driver = driver
+	}
+	return nil
+}
+
+func (c *mongodbClient) options() *options.ClientOptions {
+	clientOptions := options.Client()
+
+	if c.tlsConfig != nil {
+		clientOptions.SetTLSConfig(c.tlsConfig)
+	}
+
+	if c.replicaSetName != "" {
+		clientOptions.SetReplicaSet(c.replicaSetName)
+	}
+
+	clientOptions.ApplyURI(c.connectionString())
+	if c.timeout != 0 {
+		clientOptions.SetConnectTimeout(c.timeout)
+	}
+
+	return clientOptions
+}
+
+func (c *mongodbClient) isConnected(ctx context.Context) bool {
+	if c.driver == nil {
+		return false
+	}
+
+	if err := c.driver.Ping(ctx, nil); err != nil {
+		return false
+	}
+
+	return true
+}
+
+// for a full list of options see: https://docs.mongodb.com/manual/reference/connection-string/#connections-connection-options
+func (c *mongodbClient) connectionString() string {
+	return fmt.Sprintf("mongodb://%s%s",
+		authenticationString(c.username, c.password),
+		strings.Join(c.hosts, ","))
+}
+
+// if username based authentication is configured we should add this to connection string
+// https://docs.mongodb.com/manual/reference/connection-string/#components
+func authenticationString(username, password string) string {
```

Comment:
This looks very insecure. The documentation references a `SetAuth` function. Can we use that instead, so the client is in charge of managing these credentials?

Reply:
Yeah I like this idea 👍